File: model/src/main/java/model/organization/event/UsesEvent.java
package model.organization.event;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;

import model.organization.entity.Pmf;
import model.organization.entity.Role;
import model.organization.id.UniqueId;
import model.organization.relation.Uses;

import com.google.inject.assistedinject.Assisted;

/**
 * The {@linkplain UsesEvent} class indicates that there is an update about an {@linkplain Uses} relation.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public class UsesEvent extends AbstractEvent {

	private static final long serialVersionUID = 5013118334434822084L;
	private final UniqueId<Role> roleId;
	private final UniqueId<Pmf> pmfId;
	private transient Integer hashCode = null;
	private transient String toString = null;

	@Inject
	UsesEvent(@NotNull @Assisted final EventCategory category, @NotNull @Assisted final Uses uses) {
		super(category);
		roleId = uses.getRole().getId();
		pmfId = uses.getPmf().getId();
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Role}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Role> getRoleId() {
		return roleId;
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Pmf}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Pmf> getPmfId() {
		return pmfId;
	}

	@Override
	public boolean equals(final Object object) {
		if (object instanceof UsesEvent) {
			final UsesEvent event = (UsesEvent) object;
			return super.equals(event) && getRoleId().equals(event.getRoleId()) && getPmfId().equals(event.getPmfId());
		}
		return false;
	}

	@Override
	public int hashCode() {
		if (hashCode == null) {
			final int prime = 31;
			hashCode = super.hashCode();
			hashCode = prime * hashCode + getRoleId().hashCode();
			hashCode = prime * hashCode + getPmfId().hashCode();
		}
		return hashCode;
	}

	@Override
	public String toString() {
		if (toString == null) {
			toString = String.format("%s(%s, %s, %s)", getClass().getSimpleName(), getCategory(), getRoleId(), getPmfId());
		}
		return toString;
	}

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/relation/UsesEvent.kt
package runtimemodels.chazm.model.relation

import runtimemodels.chazm.api.entity.Pmf
import runtimemodels.chazm.api.entity.PmfId
import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.entity.RoleId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.api.relation.Uses
import runtimemodels.chazm.model.event.AbstractEvent
import runtimemodels.chazm.model.event.EventType
import runtimemodels.chazm.model.message.M
import java.util.*
import javax.inject.Inject

/**
 * The [UsesEvent] class indicates that there is an update about an [Uses] relation.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
open class UsesEvent @Inject internal constructor(
    category: EventType,
    uses: Uses
) : AbstractEvent(category) {
    /**
     * Returns a [UniqueId] that represents a [Role].
     *
     * @return a [UniqueId].
     */
    val roleId: RoleId = uses.role.id
    /**
     * Returns a [UniqueId] that represents a [Pmf].
     *
     * @return a [UniqueId].
     */
    val pmfId: PmfId = uses.pmf.id

    override fun equals(other: Any?): Boolean {
        if (other is UsesEvent) {
            return super.equals(other) && roleId == other.roleId && pmfId == other.pmfId
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(category, roleId, pmfId)

    override fun toString(): String = M.EVENT_WITH_2_IDS[super.toString(), roleId, pmfId]

    companion object {
        private const val serialVersionUID = 5013118334434822084L
    }

}
